okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java

```diff
@@ -78,8 +78,18 @@ public CallServerInterceptor(boolean forWebSocket) {
               new CountingSink(httpCodec.createRequestBody(request, contentLength));
           BufferedSink bufferedRequestBody = Okio.buffer(requestBodyOut);
 
-          request.body().writeTo(bufferedRequestBody);
-          bufferedRequestBody.close();
+          // if the requestBody is a file, this can throw an IOException
+          // which further up by the retry interceptor gets treated as a server
+          // IOException and assumes the socket is closed.
+          // We mitigate this by catching here, ensuring a close,
+          // and throwing
+          try {
+            request.body().writeTo(bufferedRequestBody);
```

Comment:
Unfortunately this will catch problems where the bufferedRequestBody stream throws a retryable IOException. But you’re on the right track. I think we need to detec _those_ inside requestBodyOut or something